File: js_api_eth/src/eth/keyring.ts
import { ethers } from "ethers";
import { verifyMessage } from "@ethersproject/wallet";
import Jazzicon from "@metamask/jazzicon";
import { erc20Abi, getProvider } from "./settings";

interface GasOptions {
  maxFeePerGas: string;
  maxPriorityFeePerGas: string;
}

/**
 * Generate a set of new mnemonic.
 * svg code from https://github.com/smart-dev817/uniswap-interface/blob/1cc5ceec4dbe29cad3378eb321fdc88e3eb79304/src/components/Identicon/index.tsx#L15
 */
async function gen(mnemonic: string, index: number) {
  const derivePath = ethers.utils.defaultPath.split("/");
  const path = derivePath.slice(0, derivePath.length - 1).join("/") + "/" + (index || 0).toString();
  const wallet = !!mnemonic ? ethers.Wallet.fromMnemonic(mnemonic, path) : ethers.Wallet.createRandom({ path });
  const icon = Jazzicon(16, parseInt(wallet.address.slice(2, 10), 16));
  return {
    mnemonic: wallet.mnemonic.phrase,
    path: wallet.mnemonic.path,
    address: wallet.address,
    svg: icon.innerHTML,
  };
}

/**
 * get address and avatar from mnemonic.
 */
async function addressFromMnemonic(mnemonic: string, derivePath: string) {
  try {
    const wallet = ethers.Wallet.fromMnemonic(mnemonic, derivePath);
    const icon = Jazzicon(16, parseInt(wallet.address.slice(2, 10), 16));
    return {
      address: wallet.address,
      svg: icon.innerHTML,
    };
  } catch (err) {
    return { error: err.message };
  }
}

/**
 * get address and avatar from privateKey.
 */
async function addressFromPrivateKey(privateKey: string) {
  try {
    const wallet = new ethers.Wallet(privateKey);
    const icon = Jazzicon(16, parseInt(wallet.address.slice(2, 10), 16));
    return {
      address: wallet.address,
      svg: icon.innerHTML,
    };
  } catch (err) {
    return { error: err.message };
  }
}

/**
 * Import keyPair from mnemonic, privateKey or keystore.
 */
async function recover(keyType: string, key: string, derivePath: string, password: string) {
  let keyPair: ethers.Wallet;
  let mnemonic = "";
  let privateKey = "";
  try {
    switch (keyType) {
      case "mnemonic":
        keyPair = ethers.Wallet.fromMnemonic(key, derivePath);
        mnemonic = key;
        break;
      case "privateKey":
        keyPair = new ethers.Wallet(key);
        privateKey = key;
        break;
      case "keystore":
        keyPair = await ethers.Wallet.fromEncryptedJson(key, password);
        break;
    }
  } catch (err) {
    return { error: err.message };
  }
  if (keyPair.address) {
    const res = await keyPair.encrypt(password, { scrypt: { N: 1 << 14 } });
    return {
      pubKey: keyPair.publicKey,
      address: keyPair.address,
      mnemonic,
      privateKey,
      keystore: res,
    };
  }
  return null;
}

/**
 * check password of an account.
 */
async function checkPassword(keystore: object, pass: string) {
  try {
    const wallet = await ethers.Wallet.fromEncryptedJson(JSON.stringify(keystore), pass);
    return { success: !!wallet.address };
  } catch (err) {
    return { success: false, error: err.message };
  }
}

/**
 * change password of an account.
 */
async function changePassword(keystore: object, passOld: string, passNew: string) {
  try {
    const keyPair = await ethers.Wallet.fromEncryptedJson(JSON.stringify(keystore), passOld);
    if (keyPair.address) {
      const res = await keyPair.encrypt(passNew, { scrypt: { N: 1 << 14 } });
      return {
        pubKey: keyPair.publicKey,
        address: keyPair.address,
        keystore: res,
      };
    }
  } catch (err) {
    return { success: false, error: err.message };
  }
}

/**
 * sign message with private key of an account.
 */
async function signMessage(message: string, keystore: object, pass: string) {
  try {
    const keyPair = await ethers.Wallet.fromEncryptedJson(JSON.stringify(keystore), pass);
    if (keyPair.address) {
      const res = await keyPair.signMessage(message);
      return {
        pubKey: keyPair.publicKey,
        address: keyPair.address,
        signature: res,
      };
    }
  } catch (err) {
    return { success: false, error: err.message };
  }
}

/**
 * get signer of a signature. so we can verify the signer.
 */
async function verifySignature(message: string, signature: string) {
  try {
    const res = verifyMessage(message, signature);
    return { signer: res };
  } catch (err) {
    return { error: err.message };
  }
}

// async function estimateTransferGas(token: string, amount: number, to: string) {
//   try {
//     if (token === "ETH") {
//       const gas = await getProvider().estimateGas({
//         to,
//         value: ethers.utils.parseEther(amount.toString()),
//       });
//       return gas.toString();
//     } else {
//       const contract = new ethers.Contract(token, erc20Abi, getProvider());
//       const decimals = await contract.decimals();
//       const gas = await contract.estimateGas.transfer(to, ethers.utils.parseUnits(amount.toString(), decimals));
//       return gas.toString();
//     }
//   } catch (err) {
//     return { success: false, error: err.message };
//   }
// }

async function transfer(token: string, amount: number, to: string, keystore: object, pass: string, gasOptions: GasOptions) {
  try {
    const keyPair = await ethers.Wallet.fromEncryptedJson(JSON.stringify(keystore), pass);
    if (keyPair.address) {
      const signer = keyPair.connect(getProvider());
      const options = {
        maxFeePerGas: ethers.utils.parseUnits(gasOptions.maxFeePerGas, 9),
        maxPriorityFeePerGas: ethers.utils.parseUnits(gasOptions.maxPriorityFeePerGas, 9),
      };
      let res: ethers.providers.TransactionResponse;
      if (token === "ETH") {
        res = await signer.sendTransaction({
          to,
          value: ethers.utils.parseEther(amount.toString()),
          gasLimit: ethers.utils.parseUnits("21", 3),
          ...options,
        });
      } else {
        const contract = new ethers.Contract(token, erc20Abi, signer);
        const decimals = await contract.decimals();
        res = await contract.transfer(to, ethers.utils.parseUnits(amount.toString(), decimals), {
          ...options,
          gasLimit: ethers.utils.parseUnits("200", 3),
        });
      }
      return {
        pubKey: keyPair.publicKey,
        address: keyPair.address,
        hash: res.hash,
      };
    }
  } catch (err) {
    return { success: false, error: err.message };
  }
}

async function signAndSendTx(tx: ethers.providers.TransactionRequest, keystore: object, pass: string, gasOptions: GasOptions) {
  try {
    const keyPair = await ethers.Wallet.fromEncryptedJson(JSON.stringify(keystore), pass);
    if (keyPair.address) {
      const signer = keyPair.connect(getProvider());
      const res = await signer.sendTransaction({
        ...tx,
        maxFeePerGas: ethers.utils.parseUnits(gasOptions.maxFeePerGas, 9),
        maxPriorityFeePerGas: ethers.utils.parseUnits(gasOptions.maxPriorityFeePerGas, 9),
      });
      return {
        pubKey: keyPair.publicKey,
        address: keyPair.address,
        hash: res.hash,
      };
    }
  } catch (err) {
    return { success: false, error: err.message };
  }
}

export default {
  gen,
  addressFromMnemonic,
  addressFromPrivateKey,
  recover,
  checkPassword,
  changePassword,
  signMessage,
  verifySignature,
  // estimateTransferGas,
  transfer,
  signAndSendTx,
};

File: lib/eth/keyring.dart
import 'dart:convert';
import 'package:polkawallet_sdk/api/apiETHKeyring.dart';
import 'package:polkawallet_sdk/service/index.dart';
import 'package:polkawallet_sdk/storage/types/GenerateMnemonicData.dart';

class ETHServiceKeyring {
  ETHServiceKeyring(this.serviceRoot);

  final SubstrateService serviceRoot;

  /// Generate a set of new mnemonic.
  Future<GenerateMnemonicData> generateMnemonic(
      {int? index, String? mnemonic}) async {
    final dynamic acc = await serviceRoot.webView!
        .evalJavascript('eth.keyring.gen("$mnemonic",$index)');
    return GenerateMnemonicData.fromJson(acc);
  }

  /// get address and avatar from mnemonic.
  Future<dynamic> addressFromMnemonic(
      {required String derivePath, required String mnemonic}) async {
    print('eth.keyring.addressFromMnemonic("$mnemonic","$derivePath")');
    final dynamic acc = await serviceRoot.webView!.evalJavascript(
        'eth.keyring.addressFromMnemonic("$mnemonic","$derivePath")');
    return acc;
  }

  /// get address and avatar from privateKey.privateKey: string
  Future<dynamic> addressFromPrivateKey({required String privateKey}) async {
    final dynamic acc = await serviceRoot.webView!
        .evalJavascript('eth.keyring.addressFromPrivateKey("$privateKey")');
    return acc;
  }

  /// Import keyPair from mnemonic, privateKey or keystore.
  /// keyType: string, key: string, derivePath: string, password: string
  Future<dynamic> importAccount({
    required ETH_KeyType keyType,
    required String key,
    required String derivePath,
    required String name,
    required String password,
  }) async {
    // generate json from js-api
    final String type = keyType.toString().split('.')[1];
    String code =
        'eth.keyring.recover("$type", "$key", "$derivePath", "$password")';
    code = code.replaceAll(RegExp(r'\t|\n|\r'), '');
    dynamic acc = await serviceRoot.webView!.evalJavascript(code);
    acc["name"] = name;

    return acc;
  }

  /// check password of account
  Future<bool> checkPassword(
      {required String keystore, required String pass}) async {
    keystore = keystore.replaceAll("\"", "\\\"");
    final res = await serviceRoot.webView!
        .evalJavascript('eth.keyring.checkPassword("$keystore", "$pass")');
    //An error Message is displayed if it fails :{ success: false, error: err.message }
    return res["success"];
  }

  /// change password of account
  Future<Map?> changePassword(
      {required String keystore,
      required String passOld,
      required String passNew}) async {
    keystore = keystore.replaceAll("\"", "\\\"");
    final res = await serviceRoot.webView!.evalJavascript(
        'eth.keyring.changePassword("$keystore", "$passOld", "$passNew")');
    return res;
  }

  /// sign message with private key of an account.
  Future<dynamic> signMessage(
      {required String message,
      required String keystore,
      required String pass}) async {
    keystore = keystore.replaceAll("\"", "\\\"");
    final res = await serviceRoot.webView!.evalJavascript(
        'eth.keyring.signMessage("$message", "$keystore", "$pass")');
    return res;
  }

  /// get signer of a signature. so we can verify the signer.
  Future<dynamic> verifySignature(
      {required String message, required String signature}) async {
    final res = await serviceRoot.webView!.evalJavascript(
        'eth.keyring.verifySignature("$message", "$signature")');
    return res;
  }

  Future<List?> getPubKeyIconsMap(List<String?> pubKeys) async {
    return await serviceRoot.eth.account.getAddressIcons(pubKeys);
  }

  Map updateKeyPairMetaData(Map acc, String? name) {
    acc['name'] = name;
    // acc['meta']['name'] = name;
    // if (acc['meta']['whenCreated'] == null) {
    //   acc['meta']['whenCreated'] = DateTime.now().millisecondsSinceEpoch;
    // }
    // acc['meta']['whenEdited'] = DateTime.now().millisecondsSinceEpoch;
    return acc;
  }
}
